frontend/src/app/testQueue/TestTimer.test.tsx

```diff
@@ -53,6 +68,102 @@ describe("TestTimer", () => {
   });
 });
 
+describe("TestTimerWidget", () => {
+  describe("telemetry", () => {
+    let context: TimerTrackEventMetadata;
+
+    beforeEach(() => {
+      context = {
+        patientId: "patient-id",
+        organizationName: "Organization",
+        facilityName: "Facility",
+        testOrderId: "test-order-id",
+      };
+    });
+
+    afterEach(() => removeTimer("internal-id"));
+
+    it("tracks a custom event when the timer is started", async () => {
+      const { appInsights } = require("../TelemetryService");
+
+      render(<DummyTestTimer testLength={15} context={context} />);
+
+      const startTimer = await screen.findByRole("button");
+
+      act(() => {
+        fireEvent.click(startTimer);
+      });
+
+      expect(appInsights?.trackEvent).toHaveBeenCalled();
+      expect(appInsights?.trackEvent).toHaveBeenCalledTimes(1);
+      expect(appInsights?.trackEvent).toHaveBeenCalledWith(
+        { name: "Test timer started" },
+        context
+      );
+    });
+
+    it("tracks a custom event when the timer reset", async () => {
+      const { appInsights } = require("../TelemetryService");
+
+      render(<DummyTestTimer testLength={15} context={context} />);
+
+      const timerButton = await screen.findByRole("button");
+
+      // Start timer
+      act(() => {
+        fireEvent.click(timerButton);
+      });
+
+      // The timer does not enter the countdown state instantly, so clicking the
+      // button in rapid succession will register as two "start timer" events.
+      // Force the timer forward -- otherwise we would need a 1sec timeout here
+      findTimer("internal-id")?.tick(Date.now());
+
+      // Reset timer
+      act(() => {
+        fireEvent.click(timerButton);
+      });
+
+      expect(appInsights?.trackEvent).toHaveBeenCalledWith(
+        { name: "Test timer reset" },
+        context
+      );
+    });
+
+    it("tracks a custom event when the timer reaches zero", async () => {
+      const { appInsights } = require("../TelemetryService");
+
+      render(<DummyTestTimer testLength={0} context={context} />);
+
+      const timerButton = await screen.findByRole("button");
+
+      act(() => {
+        fireEvent.click(timerButton);
+      });
+
+      // This is a 0-second timer, but it takes ~1 second to enter the
```

Comment:
**question (non-blocking):** This is a weird observation – any idea why it happens like this?

Reply:
I'm not totally sure - the timer tracks a `startedAt` timestamp internally that initializes to 0, and when that value is >0 then the widget considers the timer to be in its countdown state. It seems like the widget takes a full second to "realize" that it's running and re-render with the countdown active.

It looks weirder on a 0-second timer than it does on a 15-minute one 🙂 


https://user-images.githubusercontent.com/27730981/133682597-c6e94666-ba46-4b4a-97ac-931a0976239d.mov